src/hooks/useSound.ts

```diff
@@ -0,0 +1,370 @@
+  );
+  const [state, dispatch] = useReducer(reducer, {
+    ...initialState,
+    isLoop: loop || false,
+    volume,
+    playbackRate,
+  });
+
+  const audioRef = useRef<HTMLAudioElement | null>(null);
+
+  const safeAudioHelper = (action: SafeAudioAction) => {
+    if (audioRef.current) {
+      action(audioRef.current);
+    }
+  };
+
+  const cleanUpAudioRef = () => {
+    if (audioRef.current) {
+      audioRef.current.pause();
+      audioRef.current = null;
+    }
+  };
+
+  useEffect(() => {
+    audioRef.current = new Audio(url);
+
+    return () => {
+      cleanUpAudioRef();
+    };
+  }, [url]);
+
+  useEffect(() => {
+    if (!audioRef.current) {
+      return;
+    }
+
+    const cleanup = setupEventHandlers(
+      audioRef.current,
+      state.isLoop,
+      dispatch,
+      onLoad,
+      onPlay,
+      onPause,
+      onEnd
+    );
+
+    return () => {
+      cleanup();
+      cleanUpAudioRef();
+    };
+  }, [onLoad, onPlay, onPause, onEnd, state.isLoop]);
+
+  useEffect(() => {
+    safeAudioHelper((audioRef) => {
+      setupAudioElement(audioRef, state);
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state.volume, state.playbackRate, state.isLoop, state.isMuted]);
+
+  const play = () =>
+    safeAudioHelper((audioRef) => {
+      dispatch({ type: ACTION_TYPES.SET_IS_PLAYING, payload: true });
+      dispatch({ type: ACTION_TYPES.SET_IS_PAUSED, payload: false });
+      audioRef.play();
+    });
+
+  const pause = () =>
+    safeAudioHelper((audioRef) => {
+      dispatch({ type: ACTION_TYPES.SET_IS_PAUSED, payload: true });
+      audioRef.pause();
+    });
+
+  const stop = () =>
+    safeAudioHelper((audioRef) => {
+      dispatch({ type: ACTION_TYPES.SET_IS_PLAYING, payload: false });
+      dispatch({ type: ACTION_TYPES.SET_IS_PAUSED, payload: false });
+      dispatch({ type: ACTION_TYPES.SET_CURRENT_TIME, payload: 0 });
+      audioRef.pause();
+    });
+
+  const setPosition = (time: number) => {
+    safeAudioHelper((audioRef) => {
+      dispatch({ type: ACTION_TYPES.SET_CURRENT_TIME, payload: time });
+      audioRef.currentTime = time;
+    });
+  };
+
+  const setMute = () => {
+    dispatch({ type: ACTION_TYPES.SET_IS_MUTED, payload: true });
+  };
+
+  const setUnmute = () => {
+    dispatch({ type: ACTION_TYPES.SET_IS_MUTED, payload: false });
+  };
+
+  return {
+    ...state,
+    play,
+    pause,
+    stop,
+    setMute,
+    setUnmute,
+    setPosition,
+    setIsLoop: (isLoop) =>
+      dispatch({ type: ACTION_TYPES.SET_IS_LOOP, payload: isLoop }),
+    setVolume: (volume) =>
+      dispatch({
+        type: ACTION_TYPES.SET_VOLUME,
+        payload: setValidRange(0, 1, volume, 'volume'),
+      }),
+    setPlaybackRate: (rate) =>
+      dispatch({
+        type: ACTION_TYPES.SET_PLAYBACK_RATE,
+        payload: setValidRange(0.5, 4, rate, 'playbackRate'),
+      }),
+  };
+}
+
+export default useSound;
+
+// 타입 및 초기값 정의
+type AudioState = {
```

Comment:
```suggestion
interface AudioState = {
```
기본으로 interface를 사용하기로 했으니 수정해 볼까요?🤗